File: backend/models/membership.py
"""
Membership model - represents user's subscriptions and memberships
"""
from dataclasses import dataclass, field
from typing import List, Dict
from datetime import datetime


@dataclass
class Membership:
    """
    Represents a user's membership or subscription
    
    Attributes:
        id: Unique membership identifier
        name: Name of membership (e.g., "IEEE", "IEI", "American Express Platinum")
        category: Type of membership (professional, credit_card, lifestyle)
        user_id: Associated user
        start_date: When membership started
        end_date: When membership expires
        is_active: Whether membership is currently valid
        associated_benefits: List of benefit IDs linked to this membership
    """
    id: str
    name: str
    category: str  # "professional", "credit_card", "lifestyle"
    user_id: str
    start_date: datetime
    end_date: datetime
    is_active: bool = True
    associated_benefits: List[str] = field(default_factory=list)
    
    def is_expired(self) -> bool:
        """Check if membership has expired"""
        return datetime.now() > self.end_date
    
    def days_until_expiry(self) -> int:
        """Calculate days remaining until expiration"""
        delta = self.end_date - datetime.now()
        return delta.days

File: flutter_app/lib/models/membership.dart
class Membership {
  final String id;
  final String name;
  final String category;
  final String description;
  final List<String> associatedBenefits;
  final DateTime? startDate;
  final DateTime? endDate;
  final bool isActive;

  const Membership({
    required this.id,
    required this.name,
    required this.category,
    required this.description,
    required this.associatedBenefits,
    this.startDate,
    this.endDate,
    this.isActive = true,
  });

  factory Membership.fromJson(Map<String, dynamic> json) {
    return Membership(
      id: json['id'] as String,
      name: json['name'] as String,
      category: json['category'] as String,
      description: json['description'] as String,
      associatedBenefits:
          List<String>.from(json['associated_benefits'] ?? []),
      startDate: json['start_date'] != null
          ? DateTime.parse(json['start_date'] as String)
          : null,
      endDate: json['end_date'] != null
          ? DateTime.parse(json['end_date'] as String)
          : null,
      isActive: json['is_active'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'category': category,
        'description': description,
        'associated_benefits': associatedBenefits,
        'is_active': isActive,
      };

  bool get isExpired =>
      endDate != null && endDate!.isBefore(DateTime.now());

  int? get daysUntilExpiry =>
      endDate?.difference(DateTime.now()).inDays;
}
